### ISSUE TITLE
Feature request: tcp output with HA/LB support

### ISSUE THREAD

**Issue by jay7x** [rsyslog/rsyslog#3340]
It'd be great to have TCP output which able to distribute messages over multiple TCP upstreams using some logic (round-robin e.g.).

Let's imagine following setup. We have multiple API servers. Rsyslog is running on every host. It's collecting logs from API running on same host and forwarding logs via syslog-over-TCP to central log collector.

To make log collector high available right now we need to configure keepalived on multiple servers and use that virtual IP in rsyslog's `omfwd` module. But in this case it will be active-backup HA. We can't spread load over log collectors. It's possible to configure multiple `omfwd` and use `call_indirect` e.g. to distribute load over collectors but there will be no HA. Any dead upstream will cause queue overflow which may stop processing on upper ruleset or start dropping messages.

Ideally we may have `omfwd`-like output which can accept multiple upstream servers and spread messages over available uplinks using round-robin logic in simple case.

In my case I had 7 Graylog servers with TCP syslog inputs and single log collector. I'd like to distribute messages over those 7 Graylogs having both HA and LB. But there was no way so I added Kafka in-between (which is overkill a bit because both rsyslog and Graylog have queues).

Actually I'm not sure should it be `action` or maybe `ruleset` option or even generic `action` option... At time of writing I realised that ruleset may do round-robin distribution over contained actions as well.
So looking for more comments and ideas here :)

**Comment by davidelang** [comment]
This is what the rebindinterval is for. It still only sends to one destination 
at a time, but you can have it reconnect every 100 or 1000 messages, which 
allows a load balancer to spread the load across different servers. If you set 
this so that it switches up to a few times per second, the performance impact of 
the disconnect is very low, and the effect will be pretty much the same as a 
full round-robin.

I did this with Linux CLUSTERIP

see: 
https://www.usenix.org/conference/lisa12/technical-sessions/presentation/lang_david

David Lang


  On Mon, 10 Dec 2018, Yury 
Bushmelev wrote:

> Date: Mon, 10 Dec 2018 05:43:56 -0800
> From: Yury Bushmelev <notifications@github.com>
> Reply-To: rsyslog/rsyslog
>     <reply+0021b24f39e9c2fba0ba005cb5058fec6d432a6bc9aea00e92cf0000000118262f9
>     c92a169ce17345c13@reply.github.com>
> To: rsyslog/rsyslog <rsyslog@noreply.github.com>
> Cc: Subscribed <subscribed@noreply.github.com>
> Subject: [rsyslog/rsyslog] Feature request: tcp output with HA/LB support
>     (#3340)
> 
> It'd be great to have TCP output which able to distribute messages over multiple TCP upstreams using some logic (round-robin e.g.).
>
> Let's imagine following setup. We have multiple API servers. Rsyslog is running on every host. It's collecting logs from API running on same host and forwarding logs via syslog-over-TCP to central log collector.
>
> To make log collector high available right now we need to configure keepalived on multiple servers and use that virtual IP in rsyslog's `omfwd` module. But in this case it will be active-backup HA. We can't spread load over log collectors. It's possible to configure multiple `omfwd` and use `call_indirect` e.g. to distribute load over collectors but there will be no HA. Any dead upstream will cause queue overflow which may stop processing on upper ruleset or start dropping messages.
>
> Ideally we may have `omfwd`-like output which can accept multiple upstream servers and spread messages over available uplinks using round-robin logic in simple case.
>
> In my case I had 7 Graylog servers with TCP syslog inputs and single log collector. I'd like to distribute messages over those 7 Graylogs having both HA and LB. But there was no way so I added Kafka in-between (which is overkill a bit because both rsyslog and Graylog have queues).
>
> Actually I'm not sure should it be `action` or maybe `ruleset` option or even generic `action` option... At time of writing I realised that ruleset may do round-robin distribution over contained actions as well.
> So looking for more comments and ideas here :)
>
>

**Comment by jay7x** [comment]
Just got another case when this feature will help. We have legacy graylog cluster of 3 nodes. There are some clients writing syslog to global graylog inputs. Clients are using round-robin DNS to connect to syslog tcp input on graylog. After maintenance when all graylog nodes are restarted one-by-one all clients usually connected to first node restarted because it was available for longer time. This causing node overload which is kind of expected. I'd like to put rsyslog in front of graylog syslog inputs to load balance it (and for easier maintenance/debugging) but see no easy way to do.

**Comment by davidelang** [comment]
I did a presentation on how you can load balance at LISA '12 
https://www.usenix.org/conference/lisa12/technical-sessions/presentation/lang_david

**Comment by jay7x** [comment]
@davidelang thank you for that recording! 
Don't understand me wrong. I'm able to do workaround. But I'd prefer to have built-in way to distribute load across upstreams. We're not in 90s anymore. I have no real servers anymore. I'm still lucky as I have virtual machines in cloud at least. But I can't just implement routing in cloud in the way I'd like to. I can't change configuration of cloud-managed k8s. So I'd like to have something out of the box. I see this feature very promising in modern world. Elsewhere I'll start writing some syslog balancing proxy, which will implement 50% of rsyslog `imptcp` and `omfwd` in the end. I see this as plain waste of time as most parts are already there.

**Comment by davidelang** [comment]
In the cloud you have load balancers that know how many target machines you have 
(and can spin up more as needed)

point the output of rsyslog to the load balancer and have it balance across your 
targets.

It is not reasonable to have rsyslog implement a load balancer that can tie in 
to every cloud provider to know what targets exist, check their status, etc. So 
Rsyslog implements the rebind interval so that it gives the load balancer 
(whatever it is) a chance to rebalance the load. So when you first start up, 
everything may end up pointing at the first destination node to come up, but 
over a very short time, as the rebinds happen, the load balancer gets a chance 
to spread the load over all of the destination.

David Lang

**Comment by jay7x** [comment]
I'm still thinking `rebindInterval` is kind of ugly workaround. But I see you point, thank you! I'll think how to deal with it.

**Comment by davidelang** [comment]
On Thu, 14 Mar 2019, Yury Bushmelev wrote:

> I'm still thinking `rebindInterval` is kind of ugly workaround.

In many ways it is, but it's also the minimum capability needed to make other 
load balancing able to adapt to changes.

Given how TCP connections will loose messages in flight when the connection is 
cut, I don't like the idea of keeping many connections open (and vulnerable to 
being cut and loosing messages).

It may make sense to have relp support multiple connections in parallel, since 
it can track messages that failed to be delivered to one system and re-send them 
to another, but it's a lot of extra complexity in the relp module.

David Lang

**Comment by jay7x** [comment]
> It may make sense to have relp support multiple connections in parallel, since it can track messages that failed to be delivered to one system and re-send them to another, but it's a lot of extra complexity in the relp module.

So what I'm proposing is actually to have way to distribute messages across some kind of resource pool (actions in ruleset). When implemented on this level it will support any outputs including `omrelp`.

**Comment by davidelang** [comment]
On Thu, 14 Mar 2019, Yury Bushmelev wrote:

>> It may make sense to have relp support multiple connections in parallel, since it can track messages that failed to be delivered to one system and re-send them to another, but it's a lot of extra complexity in the relp module.
>
> So what I'm proposing is actually to have way to distribute messages across some kind of resource pool (actions in ruleset). When implemented on this level it will support any outputs including `omrelp`.

there is an ability to do 'call $.var', so if you can calculate a hash on the 
message and key from that it could work.

We do have global variables ($\), but they are very expensive to use

David Lang

**Comment by jay7x** [comment]
That will fail if action fail in ruleset. Can have N rulesets with same M actions though.. should work but looks like maintenance hell if you writing config manually.

**Comment by gecube** [comment]
I believe that this feature is a must, but need more investigation and research how to be implemented. I'll explain. Now there is great abyss between classical simple Unix toolings (which realize some simple functionality, but very good) and big combines (like Zabbix - a number of integrations and very complex internal structure). Also there is tendency to make systems distributed (for example, Kafka, kubernetes etc.). So correctly implementing balancing and HA in rsyslog we can stay in the same rsyslog stack without necessity of using external entities, particularly in cloud environments (which are dynamic by their nature).
So - +1

**Comment by davidelang** [comment]
Implementing Load Balancing can be a very complex thing to do, there are a lot 
of different load balancing strateges to implement. Rsyslog (like most 
opensource projects) is always short on manpower, so we have to pick where to 
spend our efforts (we also welcome contributions on things that we don't spend 
our efforts on)

Rsyslog had a feature to allow it to support existing load balancers, you can 
configure it (at least for TCP, UDP and elasticsearch outputs, I think I've seen 
it in a few other outputs) to close and re-open the connection every X messages, 
which allows non-rsyslog load balancing solutions to redirect the next set of 
logs to a different destination.

The non-rsyslog load balancing can be an appliance (F5 or similar), or it can be 
software such as CLUSTERIP on Linux (as I describe in the paper I presented at 
LISA '12 
https://www.usenix.org/conference/lisa12/technical-sessions/presentation/lang_david 
)

This seems to be the minimum viable feature to be able to support load 
balancing (And many HA scenarios)

For HA, we also support that action-only-if-the-last-action-failed to impelement 
simple 'send to server A unless it's down, then send to server B' functionality.

So I believe that we have sufficient support to allow the admin to implement 
multiple types of HA/LB, and the next step of implementing specific HA/LB 
strategies is a big one, supporting some but not all is going to generate lots 
of requests to implement more of the strategies, and I don't think that's a good 
use of the core developers time.

Again, we'd be happy to accept contributions that implement the HA/LB 
strategies. But it may mean that when people ask us to implement more 
strategies or ask for support troubleshooting the functionality, we may have to 
answer that it's a contributed feature that we can only support so far.

David Lang

**Comment by rgerhards** [comment]
Apologies for the very late reply. This is probably no longer relevant for the original reporter, but I am adding the answer for people who find this issue through search. We do not lightly close old issues just because of age; we are working through the backlog and closing them when the requested capability has been implemented or the thread has a clear resolution.

The original request was for `omfwd` to support forwarding to multiple TCP upstreams with load distribution and failover rather than relying only on an external load balancer plus `RebindInterval`. Current rsyslog has moved in that direction: `omfwd` now supports target arrays and connection-pool style forwarding, with parameters such as `target=[...]`, matching `port=[...]` / `protocol=[...]`, `pool.name`, `pool.size`, and `pool.resumeInterval` for TCP destinations. This is the modern built-in path for a compact multi-target forwarding setup.

The older advice in the thread is still relevant for some deployments: `rebindInterval` remains useful when an external L4/load-balancer should make the distribution decision, and broader HA/LB policy across arbitrary action types is still a larger design topic. But the specific `omfwd` multi-target TCP output capability requested here is now present, so I am closing this as implemented/superseded. If a particular load-balancing strategy is still missing from current `omfwd` pools, please open a new focused issue for that strategy and expected failure/retry semantics.

**Comment by gecube** [comment]
@rgerhards Hey! It is very nice that you put your comment here, so we don't feel abandoned. And I am very happy that the project is still actively developing. Thank you very much!

**Comment by rgerhards** [comment]
@gecube thx for you kind words, much appreciated. And, yes, there were some really bad times. Full story is here: https://rainer.gerhards.net/2025/08/sluggish-responses-and-how-we-plan-to-do-better.html

Bottom line: rsyslog is more alive than ever and evolving very rapidly. Feature suggestions are very welcome.

**Comment by jay7x** [comment]
Actually, I'm going to use rsyslog as a log collector & shipper very soon again, so this is still quite relevant :)

**Comment by rgerhards** [comment]
@jay7x 

> Actually, I'm going to use rsyslog as a log collector & shipper very soon again, so this is still quite relevant :)

great to hear! Again, pls let us know everything that might be useful for you. Also glitches (code and doc), missing capabilities, potential bugs - we are better prepared than ever to find solutions fast (and I hope I will not mess up after that strong wording ;-)).

**Comment by jay7x** [comment]
@rgerhards off topic here, but I asked about Debian support in Github discussions while ago.. that'd be very useful to get the answer there :)

**Comment by rgerhards** [comment]
thx for the OT ping!
